src/charm.py

```diff
@@ -1,181 +1,93 @@
-        self.model.pod.set_spec(
-            {
-                "version": 3,
-                "containers": [
-                    {
-                        "name": "mlmd",
-                        "command": ["/bin/metadata_store_server"],
-                        "args": args,
-                        "imageDetails": image_details,
-                        "ports": [
-                            {
-                                "name": "grpc-api",
-                                "containerPort": int(self.model.config["port"]),
-                            },
-                        ],
-                        "volumeConfig": volumes,
-                        "kubernetes": {
-                            "livenessProbe": {
-                                "tcpSocket": {"port": "grpc-api"},
-                                "initialDelaySeconds": 3,
-                                "periodSeconds": 5,
-                                "timeoutSeconds": 2,
-                            },
-                            "readinessProbe": {
-                                "tcpSocket": {"port": "grpc-api"},
-                                "initialDelaySeconds": 3,
-                                "periodSeconds": 5,
-                                "timeoutSeconds": 2,
-                            },
-                        },
-                    }
+        # KubernetesServiceInfoProvider for broadcasting the GRPC service information
+        self._svc_port = self.config["grpc-port"]
+        self._k8s_svc_info_provider = KubernetesServiceInfoProvider(
+            charm=self, relation_name=RELATION_NAME, name=GRPC_SVC_NAME, port=self._svc_port
+        )
+
+        # Charm logic
+        self.charm_reconciler = CharmReconciler(self)
+
+        self.leadership_gate = self.charm_reconciler.add(
+            component=LeadershipGateComponent(
+                charm=self,
+                name="leadership-gate",
+            ),
+            depends_on=[],
+        )
+
+        self.kubernetes_resources = self.charm_reconciler.add(
+            component=KubernetesComponent(
+                charm=self,
+                name="kubernetes:svc",
+                resource_templates=K8S_RESOURCE_FILES,
+                krh_resource_types={Service},
+                krh_labels=create_charm_default_labels(
+                    self.app.name, self.model.name, scope="svc"
+                ),
+                context_callable=lambda: {
+                    "app_name": self.app.name,
+                    "namespace": self.model.name,
+                    "grpc_port": self._svc_port,
+                },
+                lightkube_client=lightkube.Client(),
+            ),
+            depends_on=[self.leadership_gate],
+        )
+
+        self.mlmd_container = self.charm_reconciler.add(
+            component=MlmdPebbleService(
+                charm=self,
+                name="mlmd-grpc-service",
+                container_name="mlmd-grpc-server",
```

Comment:
we should probably leave the container name the same as before `mlmd` to prevent issues during upgrade like we've seen in kfp https://github.com/canonical/kfp-operators/issues/365

Reply:
I believe that issue was caused when changing the name of a sidecar container, since this is a podspec -> sidecar re-write, I don't think that'll affect us. I can test the upgrade path as well, but I don't think this will be a problem.

Reply:
that's true because currently an upgrade path for this charm would be to remove and re-deploy it due to the same issue we saw in https://github.com/canonical/bundle-kubeflow/issues/732

Reply:
The error message will be slightly different, but yes, we need to remove the podspec charm and deploy a new one.

```
ERROR Juju on containers does not support updating storage on a statefulset.
The new charm's metadata contains updated storage declarations.
You'll need to deploy a new charm rather than upgrading if you need this change.
 not supported (not supported)
```

Even if we did not have to do this, the container name won't be an issue because we are not creating a new container in the same Pod, which will not result in the issue that https://github.com/canonical/kfp-operators/issues/365 mentions.

I think then we can keep the name as is, wdyt?